### PR TITLE
Split #3054: Add base classes [ready]

### PR DESCRIPTION
**Pull request by polybassa**
Add base classes for a new and clean implementation of automotive scanners and enumerators

### CODE REVIEW

**Comment by polybassa** [comment]
@gpotter2 Would it be possible to get this PR in, so that I can continue with the split of #3054

File: scapy/contrib/automotive/scanner/test_case.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+if TYPE_CHECKING:
+    from scapy.contrib.automotive.scanner.configuration import AutomotiveTestCaseExecutorConfiguration  # noqa: E501
+
+
+if six.PY34:
+    from abc import ABC, abstractmethod
+else:
+    from abc import ABCMeta, abstractmethod
+    ABC = ABCMeta('ABC', (), {})  # type: ignore
```
Comment (gpotter2) [inline scapy/contrib/automotive/scanner/test_case.py:29]: You can just do
`@six.add_metaclass(abc.ABCMeta)`
and you don't need all of that.

**Review by gpotter2 — COMMENTED** [review]
Wouldn't it make more sense if the test-related files were in the `test/` folder.
Remember that the `test/` folder isn't included in builds, so it would make scapy lighter.

**Comment by polybassa** [comment]
@gpotter2 Sorry for the confusion with the name `AutomotiveTestCase` and `AutomotiveTestCaseExecutor`.
Prior to my refactoring, I called these classes `Enumerator` and `Scanner`. But during multiple practical tests, I found these two classes very useful for further purposes. For example, I used an `Enumerator` class to validate if a certain bug is present. Therefore I came up with a more generic naming for these two classes. I changed the base class for `Enumerator` into `AutomotiveTestCase` and `Scanner` into `AutomotiveTestCaseExecutor`. The derived classes still remain the name `Enumerator` and `Scanner`. 

The `AutomotiveTestCase` class basically defines a generic interface for the `AutomotiveTestCaseExecutor`. The unique part of this two classes is the automated aggregation (reverse engineering) of a device under tests "system state" in a directed graph. Since automotive control units are very state driven, for my scanners, it's very important to identify the current system state in order to obtain reproducible and reliable results. 

Regarding the `AutomotiveTestCaseExecutor`, currently I derive `Scanner` classes from this base class. On my current experiments, I also use this class for a Fuzzer of automotive protocols with system state awareness. Also in this use case, I find it more natural to derive "AutomotiveFuzzer" and "Scanner" from "AutomotiveTestCaseExecutor". The old naming scheme would lead to a inheritance of "AutomotiveFuzzer" from "Scanner", which I find stranger. 

To summarize that.. I've chosen this names to express a broader use of these base-classes, not only for enumeration and scanning. I'm currently experimenting with more use cases and plan to release for example AutomotiveFuzzers in the future.

File: scapy/contrib/automotive/scanner/test_case.py (hunk 1 above)
Comment (polybassa) [reply scapy/contrib/automotive/scanner/test_case.py:29]: Thanks a lot for this hint

**Comment by polybassa** [comment]
@gpotter2 could this PR get merged, please? My paper about automotive protocol scanners, I've implemented with Scapy, got accepted and will be published on 12.5.2021. I would be happy, if by that time, the relevant code (#3054) got merged.

**Review by gpotter2 — APPROVED** [review]
